### PR TITLE
Replace Vec<PK> in authority quorum with equivalent bitmap

### PR DESCRIPTION
**Pull request by punwai**
Changed the authority quorum struct to store a bitmap instead of a vector of Public Keys:

```
pub struct Ed25519AuthorityQuorumSignInfo<const STRONG_THRESHOLD: bool> {
    pub epoch: EpochId,
    pub signatures: Vec<Ed25519AuthoritySignature>,
    pub signers_map: CommitteeMap // i.e. bitmap
}
```

Each index in the bitmap corresponds to an authority in the committee. A '1' in the index indicates that the authority has signed the certificate, and a '0' indicates that they have not. The most-significant bit in `signers_map` corresponds to the authority whose public key is the smallest when sorted in ascending order.

**Note: Builds off my previous PR which introduces BLS signature support**

### CODE REVIEW

**Review by huitseeker — CHANGES_REQUESTED** [review]
1. This is a obvious ACK on the direction, those signatures should use a bitmap provided we know the committee,
2. Could you rebase the change so that it does not depend on #2877 ? Adopting BLS signatures seems a much more controversial subject than what we have at hand with bitmaps.
3. [bitmaps](https://crates.io/crates/bitmaps) is nice, but an even better option for the low-level implementation of the bitmap is [roaring-rs](https://github.com/RoaringBitmap/roaring-rs). Besides its sparsity, it has a byte representation that's a polyglot standard, allowing many libraries to generate bytes compatible with our message implementations. You can see [here](https://github.com/MystenLabs/sui/pull/798/files#diff-efc836efe81761b177b47defdd4705ec343b0e898d5af78202b8b5251dce0e96) for an example of using this library somewhere in an old PR.
4. I would love to see a couple of unit tests referring specifically to the ability to re-construct the quorum of signers. In particular, I'd like to see the behavior of `new_with_signatures`,  `add_signature` and `authorities_from_bitmap` correlated, including in pathological cases (authorities not in the committee, authority reuse, different lengths betw Bitmap and committee, etc)

File: crates/sui-types/src/committee.rs
Comment (huitseeker) [inline crates/sui-types/src/committee.rs:98]: Same remark, if the index is out of bounds.

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [inline crates/sui-types/src/crypto.rs:568]:
```suggestion
    // This takes log(sig) time, do not use if not necessary
```

File: crates/sui-types/src/committee.rs
Comment (huitseeker) [inline crates/sui-types/src/committee.rs:93]: We probably want to make this return `u32`, as this will break the map beyond the `u32` limit.

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [inline crates/sui-types/src/crypto.rs:569]: Here's something that may be useful (food for thought, not for this PR): is there a batch addition of several signatures at the same time, which would perform better?

File: crates/sui-types/src/sui_serde.rs
Comment (huitseeker) [inline crates/sui-types/src/sui_serde.rs:28]: `#[inline]` would be appropriate here I think.

File: crates/sui-types/src/unit_tests/messages_tests.rs
Comment (huitseeker) [inline crates/sui-types/src/unit_tests/messages_tests.rs:158]: It would be great to split this into three tests!

File: crates/sui-types/src/sui_serde.rs
Comment (huitseeker) [inline crates/sui-types/src/sui_serde.rs:192]: Here I would emphasize and link to the [roaring bitmap on-disk format](https://github.com/RoaringBitmap/RoaringFormatSpec).

File: crates/sui-types/src/unit_tests/messages_tests.rs
Comment (huitseeker) [inline crates/sui-types/src/unit_tests/messages_tests.rs:152]: I would love to see negative testing here:
- making which orders we depend on clear, by checking an `AuthorityQuorumSignInfo` with a:
    + permuted `signatures` field,
    + permuted `Committee` order,
- We can receive any adversarial `AuthorityQuorumSignInfo` deserialized from the network, and provided by an adversary: what are the pathological cases we can receive, and can we make sure our verification functions return with a sentinel error, rather than panic out?
    + bitmap contents that don't match the signature vec,
    + empty bitmap / signature contents,

**Review by huitseeker — COMMENTED** [review]
Thanks a lot! Superb improvements everywhere.
My remark on testing is the most important comment.

File: crates/sui-types/src/committee.rs
Comment (punwai) [reply crates/sui-types/src/committee.rs:93]: Hashmap.get() returns usize, if this is the case, is this still a problem?

Comment (huitseeker) [reply crates/sui-types/src/committee.rs]: `usize` is platform dependent, and may mean `u64`. 
This is probably not the first wrong thing about a 4GB-long table of authority indexes, but one thing it will not do is work with our `u32`-bound Bitmap.

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [reply crates/sui-types/src/crypto.rs:593]: Yes, especially as you immediately error out of the result in most cases, except ... accountability is actually useful.
While we haven't implemented it yet, some verification errors should lead to signaling a problem, and returning an iterator of results gives us the flexibility to implement this later: good work!

File: crates/sui-types/src/unit_tests/messages_tests.rs
Comment (huitseeker) [inline crates/sui-types/src/unit_tests/messages_tests.rs:358]: Nit: I would prefix with `test_`

**Review by huitseeker — APPROVED** [review]
Overall excellent work! Good to go modulo fixes on a couple of somewhat verbose iterations, that I trust you to tweak before merging!

File: crates/sui-types/src/unit_tests/messages_tests.rs
Comment (punwai) [reply crates/sui-types/src/unit_tests/messages_tests.rs:161]: Yep, when we create the committees we sort authorities! :)

File: crates/sui-types/src/unit_tests/messages_tests.rs
Comment (punwai) [reply crates/sui-types/src/unit_tests/messages_tests.rs:141]: No particular reason, I just forgot to delete it after some tests.

**Pull request merged**
